rustworkx-core/src/coloring.rs

```diff
@@ -13,13 +13,90 @@
 use std::cmp::Reverse;
 use std::hash::Hash;
 
+use crate::connectivity::isolates;
 use crate::dictmap::*;
 use crate::line_graph::line_graph;
 use hashbrown::{HashMap, HashSet};
 use petgraph::graph::NodeIndex;
-use petgraph::visit::{EdgeCount, EdgeRef, IntoEdges, IntoNodeIdentifiers, NodeCount};
+use petgraph::visit::{
+    EdgeCount, EdgeRef, GraphBase, GraphProp, IntoEdges, IntoNeighborsDirected,
+    IntoNodeIdentifiers, NodeCount, NodeIndexable,
+};
+use petgraph::{Incoming, Outgoing};
 use rayon::prelude::*;
 
+/// Compute a two-coloring of a graph
+///
+/// If a two coloring is not possible for the input graph (meaning it is not
+/// bipartite), `None` is returned.
+///
+/// Arguments:
+///
+/// * `graph` - The graph to find the coloring for
+///
+/// # Example
+///
+/// ```rust
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::coloring::two_color;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = UnGraph::<i32, i32>::from_edges(&edge_list);
+/// let coloring = two_color(&graph).unwrap();
+/// let mut expected_colors = DictMap::new();
+/// expected_colors.insert(NodeIndex::new(0), 1);
+/// expected_colors.insert(NodeIndex::new(1), 0);
+/// expected_colors.insert(NodeIndex::new(2), 1);
+/// expected_colors.insert(NodeIndex::new(3), 0);
+/// expected_colors.insert(NodeIndex::new(4), 1);
+/// assert_eq!(coloring, expected_colors)
+/// ```
+pub fn two_color<G>(graph: G) -> Option<DictMap<G::NodeId, u8>>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + GraphBase
+        + GraphProp
+        + NodeCount,
+    <G as GraphBase>::NodeId: std::cmp::Eq + Hash,
+{
+    let mut colors = DictMap::with_capacity(graph.node_count());
+    for node in graph.node_identifiers() {
+        if colors.contains_key(&node) {
```

Comment:
If I understand correctly, this outer loop allows to handle disconnected graphs. Why do we need an explicit treatments for `isolates` (wouldn't they be assigned color `1`)?

Reply:
That's a fair point, we probably don't need to I did it to assign the nodes color 0, but 1 is just a valid answer for an isolated node so we should just simplify the code to make it faster.

Reply:
Done in: https://github.com/Qiskit/rustworkx/pull/1002/commits/4c85c9bc0c9404cb4722626913edd3f685f6acd3